services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/prometheus_metrics.py

```diff
@@ -0,0 +1,188 @@
+import asyncio
+import logging
+from collections import deque
+from collections.abc import Sequence
+from datetime import datetime
+from textwrap import dedent
+from typing import Final
+
+import arrow
+from fastapi import FastAPI, status
+from models_library.callbacks_mapping import CallbacksMapping, UserServiceCommand
+from pydantic import BaseModel, NonNegativeFloat, NonNegativeInt
+from servicelib.background_task import cancel_task
+from servicelib.logging_utils import log_context
+from servicelib.sequences_utils import pairwise
+from simcore_service_dynamic_sidecar.core.errors import (
+    ContainerExecContainerNotFoundError,
+)
+
+from ..models.shared_store import SharedStore
+from .container_utils import run_command_in_container
+
+_logger = logging.getLogger(__name__)
+
+_MAX_DEFAULT_METRICS_SCRAPE_INTERVAL: Final[NonNegativeFloat] = 60.0
+_MIN_ELEMENTS: Final[NonNegativeInt] = 2
+_MAX_PROMETHEUS_SAMPLES: Final[NonNegativeInt] = 5
+_TASK_CANCELLATION_TIMEOUT_S: Final[NonNegativeInt] = 2
+
+_USER_SERVICES_NOT_STARTED: Final[str] = "User service(s) was/were not started"
+_EXTRA_METRIC_NAME: Final[str] = "dy_sidecar_event_generated_at"
+_EXTRA_METRIC_DESCRIPTION: Final[str] = "last time data was updated"
+
+
+def _get_user_services_scrape_interval(
+    last_prometheus_query_times: Sequence[datetime],
+) -> NonNegativeFloat:
+    if len(last_prometheus_query_times) < _MIN_ELEMENTS:
+        return _MAX_DEFAULT_METRICS_SCRAPE_INTERVAL
+
+    time_paris: list[tuple[datetime, datetime]] = list(
```

Comment:
where is the tour eiffel?
```suggestion
    time_pairs: list[tuple[datetime, datetime]] = list(
```